### PR TITLE
Template Parts types

### PR DESCRIPTION
**Pull request by luwes**
Adds most types for the template parts utilities.

One difficult one is the `element` argument of the `parse()` function. 
If you feel the calling please go ahead, I couldn't figure it out right away.

### CODE REVIEW

File: src/js/utils/template-parts.js [hunk 1]
```diff
@@ -35,24 +36,35 @@ export const defaultProcessor = {
...
-  update(state) {
+
+  update(/** @type Record<string, any> */ state) {
     this.#processor.processCallback(this, this.#parts, state);
   }
 }
 
 // collect element parts
-export const parse = (element, parts = []) => {
+export const parse = (element, /** @type [string, Part][] */ parts = []) => {
```
Comment (cjpillsbury) [inline src/js/utils/template-parts.js:67]: suggestion(non-blocking): consider adding a `@typedef` for your `PartsTuple` and/or `PartsTupleArray` (See: https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#typedef-callback-and-param)

**Review by cjpillsbury — APPROVED** [review]
Haven't validated the types "in action," but from a code eval perspective, this looks 👌. Only a few non-blocking comments.

File: src/js/utils/template-parts.js [hunk 2]
```diff
@@ -157,33 +170,44 @@ export const tokenize = (text) => {
...
     return this.#items.length;
   }
-  item(index) {
+
+  item(/** @type number */ index) {
     return this.#items[index];
   }
-  append(...items) {
+
+  append(/** @type Array<AttrPart | string> */ ...items) {
```
[the remark this answers is not in the document]
Comment (luwes) [reply src/js/utils/template-parts.js:202]: I tried this, it errored for some reason 🤷

File: src/js/utils/template-parts.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (luwes) [reply src/js/utils/template-parts.js:67]: yes, I tried it's the difficult one I mentioned in the description. see Slack also

**Pull request merged**